Fix Stock View sometimes not showing all items

ToMe25 opened the pull request:
Closes #4758
Closes #
Closes #

# Changes proposed in this pull request:
- Fix item handling in StockView to prevent it from only showing the amount if items from one stack per Item type per rack.
- Add the Hut block inventory to the stock view.
-

Review please

ToMe25 committed:
Fix Stock View sometimes not showing all items

ToMe25 committed:
Fix formatting

Raycoms commented:
Stack has a max size of 64, and for some things even less, you'll need to store this in an itemStorage object to fix this.

Raycoms committed:
Merge branch 'version/1.15' into fix/4758

ToMe25 commented:
> Stack has a max size of 64

In theory maybe, but it does work like this as can be seen here:
![2020-04-25_22 11 24](https://user-images.githubusercontent.com/38815969/80289866-d5ee2700-8741-11ea-8326-d3064e74b21d.png)
All these Hut blocks are in the same rack.

lexustec commented:
It should count together all Stacks.. so it shows ONLY one Itemtype per Stack.. so You know exactly how much You have in a glance

It is not a chest view where You have stacks it is like the View in Refined storage

ToMe25 committed:
Move the fix to TileEntityRack

ToMe25 commented:
I moved the Fix to the TileEntityRack class as that seems to be the less hacky solution, tell me if i should moce it back.
> It should count together all Stacks

i know, it still does that, the problem before was that it only counted some stacks.

lexustec commented:
You mean it counted the Stacks twice ?

ToMe25 commented:
No.
lets say you have two stacks of cobble in one rack, then it counted only one of them.
If you however have two stacks of cobble in different racks it counts both.

Raycoms commented on src/api/java/com/minecolonies/api/tileentities/TileEntityRack.java:
```diff
@@ -225,12 +225,14 @@ public void updateItemStorage()
             }
+            stackCopy.setCount(amount);
             content.put(storage, amount);
         }
```
```
int amount = ItemStackUtils.getSize(stack);
if (content.containsKey(storage))
{
amount += content.remove(storage);
}
content.put(storage, amount);
```
Why isn't this enough?

ToMe25 replied:
because [this](https://github.com/ldtteam/minecolonies/blob/84fa0aaf06052529fb9c664f27a6f715c3b2bb2b/src/main/java/com/minecolonies/coremod/client/gui/WindowHutAllInventory.java#L150) only uses the stack, not the amount.
both in the current, and my changed version.

lexustec replied:
that should work normally

Raycoms replied:
it gets the amount from the stack and adds it to the amount in the map. That should be enough

ToMe25 replied:
Should i change it back to how it was in the second commit then? that didn't change TileEntityRack at all.

Raycoms replied:
none of this should be necessary of the above.

ToMe25 replied:
does that mean you prefer leaving this alone like it was, and just changing the WindowHutAllInventory?

Raycoms replied:
The above code is correct. It should not be the cause of any issue.

ToMe25 replied:
ok i changed it back.

lexustec commented:
Okay ,.... i tried that btw.. this is funny maybe i lost sth while having issues with my GIT

ToMe25 committed:
Move the fix back to WindowHutAllInventory

as that seems to be the prefered solution.

ToMe25 committed:
oops, formatting again

marchermans commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutAllInventory.java:
```diff
@@ -147,7 +147,9 @@ private void updateResources()
                 for (final Map.Entry<ItemStorage, Integer> entry : storage.entrySet())
                 {
-                    items.add(new ItemStorage(entry.getKey().getItemStack(), entry.getValue(), false).getItemStack());
+                	ItemStack stack = entry.getKey().getItemStack().copy();
```
This does not work officially, and even though it works in the current Mc version it is not officially supported. As such please change this to use itemstorage instances or use the proposed change by @Raycoms.

ToMe25 replied:
should i just change the items list to use ItemStorage instead of ItemStack then?

ToMe25 replied:
I did get rid of it now.

Raycoms commented:
okay, I was looking over all of this again and I think there are things we can improve:

https://github.com/ldtteam/minecolonies/blob/3f60f2e9d059d2f00f6d923b273936db8c7faba9/src/main/java/com/minecolonies/coremod/client/gui/WindowHutAllInventory.java#L169

Instead of filling a list and then filling the map in my link. We can move this map up to the beginning of the class and count up in this map from the beginning. This should also solve this issue.

ToMe25 committed:
Update this to no longer misuse ItemStacks

ToMe25 commented:
OK i will do that.

lexustec commented:
It happened after i changed the code from the original aproach with the new map   

```
if (rack instanceof TileEntityRack)
            {
                 int allSlots = ((AbstractTileEntityRack) rack).getInventory().getSlots();
                        Log.getLogger().info("Used Slots " + allSlots);
                        for (int i = 0; i < allSlots; i++)
                        {
                            if (!((AbstractTileEntityRack) rack).getInventory().getStackInSlot(i).isEmpty())
                            {
                                items.add(((AbstractTileEntityRack) rack).getInventory().getStackInSlot(i));
                            }
                        }


            }

```

that was the old aproach..

ToMe25 committed:
Remove ItemStorage List

ToMe25 commented:
@Raycoms you mean like this?

lexustec commented:
the change of allItems will break the view

ToMe25 commented:
Again if absolutely no change to TileEntityRack is wanted, i can undo that and instead get the value of the map in WindowHutAllInventory and set the amount there.

ToMe25 commented:
> the change of allItems will break the view

allItems? i am not messing with anything names allItems.

lexustec commented:
sorry i read it wrong

Raycoms commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutAllInventory.java:
```diff
@@ -135,7 +135,7 @@ private void updateResources()
 
-        final List<ItemStack> items = new ArrayList<>();
+        final Map<ItemStorage, ItemStorage> storedItems = new HashMap<>();
         final World world = building.getColony().getWorld();
```
Yes better, and make this a Map<ItemStorage, Integer> and increment the quantity in the. And keep ItemStorage the same

ToMe25 replied:
Did that.

Raycoms commented on src/api/java/com/minecolonies/api/tileentities/TileEntityRack.java:
```diff
@@ -231,6 +231,7 @@ public void updateItemStorage()
                 amount += content.remove(storage);
             }
+            storage.setAmount(amount);
             content.put(storage, amount);
```
this should not be necessary.

ToMe25 replied:
ok i will remove it.

ToMe25 replied:
I removed it.

ToMe25 committed:
Change storedItems to Map<ItemStorage, Integer>

ToMe25 commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutAllInventory.java:
```diff
@@ -158,29 +165,25 @@ else if (rack instanceof ChestTileEntity)
+        storedItems.forEach((storage, amount) -> {
+            storage.setAmount(amount);
+            filterItems.add(storage);
+        });
```
Is like this OK, or do i have to find another way to set the amount of the ItemStorage?

Raycoms replied:
Should be fine like this.
Did you test this?

ToMe25 replied:
Yes i did.

ToMe25 replied:
If you want to suggest any more changes, i will look at it tomorrow, today its late enough :)

Raycoms replied:
Looks fine for me now. Does this solve all the issues there were before?

Can you add an additional check to query the rack from the position of the building itself (it's not include din the list of containers, so it needs an additional line to add the stacks from it.

ToMe25 replied:
"Does this solve all the issues there were before?" idk, i don't know about many issues with this view, i just made this to fix the linked issue.
"Can you add an additional check to query the rack from the position of the building itself" I can do that, but i think lexustec said he is working on that, so i wouldn't put that in here.

ToMe25 replied:
"Can you add an additional check to query the rack from the position of the building itself"
I have a working version for the warehouse now, i will test it with some other buildings, and than hopefully push it.

ToMe25 committed:
Add the hut block inventory to the stock view

after some more conversation with luxustec i now made a working version
of this.

lexustec commented:
saw it ..... nice fix :-)

ToMe25 committed:
Remove unnecessary import

also rename storage map to rackStorage

Raycoms commented:
:shipit: 

[![Approved with PullApprove](https://img.shields.io/badge/maintainers-approved-brightgreen.svg)](https://pullapprove.com/ldtteam/minecolonies/pull-request/4759/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=ldtteam/minecolonies)

Raycoms committed:
Merge branch 'version/1.15' into fix/4758

Raycoms committed:
Merge branch 'version/1.15' into fix/4758

ToMe25 commented:
For some reason this currently isn't working for vanilla chests, i will work on fixing that now.

ToMe25 commented:
I looked at the problem with the chests some more, and i think there are three solutions for it, but i am not sure which one to use.
The solutions:
1. Ignore that this isn't working with chests as it didn't before this PR either
2. Create a custom packet to transfer the chest inventory to the client to read it
3. I am not sure, but maybe it would be possible to (mis?)use a vanilla packet to transfer the chest contents

Please tell me what the prefered solution would be

Raycoms commented:
That sounds weird, the chest contents should also exist on the client side, don't they?

ToMe25 commented:
I haven't found anything that is able to get them clientside without a custom packet, and the clientside tileentity doesn't contain any items even when the chest is full.

Raycoms commented:
@OrionDevelopment did you know that?

ToMe25 commented:
I also thought that is weird, but i couldn't find anything that is able to read this clientside.

Raycoms committed:
Merge branch 'version/1.15' into fix/4758

ToMe25 committed:
Merge branch 'version/1.15' of https://github.com/ldtteam/minecolonies into fix/4758

Raycoms committed:
Merge branch 'version/1.15' into fix/4758

ToMe25 committed:
Merge branch 'version/1.15' of https://github.com/ldtteam/minecolonies into fix/4758

ToMe25 committed:
Merge branch 'fix/4758' of https://github.com/ToMe25/minecolonies into fix/4758

ToMe25 committed:
Change chest handling to use LockableTileEntity instead of
ChestTileEntity

marchermans commented:
No this is the normal state. All vanilla inventories do not sync their inventory. Only when a container is opened the container contents are synced.

ToMe25 commented:
So should i just leave the chest problem be, or make a packet for that?
I don't think there is a vanilla packet for this.

Raycoms commented:
I'd say then remove the vanilla support code and leave it be. And we tell our schematic authors to not use chests at all anymore =)

ToMe25 commented:
ok, but i think no chests will be a loss for the mod, as just one type of storage just always has places where it looks wrong.

ToMe25 committed:
Remove chest support as it wasn't working anyways

ToMe25 committed:
oops, imports

lunadragnstar commented:
We primarily use racks, but sometimes we use chests and barrels in builds

ToMe25 commented:
barrels already don't fully work, i think about making a pr for that tho...
but i probably wont as this doesn't supports chests now anyways.

lunadragnstar commented:
Is this specifically for the warehouse, or for all storage

ToMe25 commented:
is what for the warehouse or all storage?

Raycoms commented:
@lunadragnstar it's for all buildings.

Raycoms commented:
@OrionDevelopment what is your opinion on this?

marchermans reviewed (approved):
Not the greatest of solutions. But alright.

Raycoms committed:
Merge branch 'version/1.15' into fix/4758
